### PR TITLE
DockerOperator: use DOCKER_HOST as default for docker_url

### PR DESCRIPTION
**Pull request by maresb**
Respect the `DOCKER_HOST` environment variable with the DockerOperator.

---

I was really surprised that the `DOCKER_HOST` environment variable is ignored. I was also surprised that I didn't find any proposal to add it, so apologies in case I missed a previous discussion. Might this be an acceptable change?

Of course this will break in situations where  `DOCKER_HOST` is set, but the user is relying on the default value of `docker_url` to override this setting. Having such a setup is really begging for trouble, so I wouldn't feel so guilty about breaking those cases.

TODO if I get :+1: to proceed:

- [x] Add tests (any suggestions?)
- [x] In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).
- [x] Anything else?

### CODE REVIEW

**Comment by Taragolis** [comment]
> Add tests (any suggestions?)

Simple tests which check respect docker host, case if env set to empty string (edge case) and if DOCKER_HOST not set

Tests usual stored into the test/{path-from-airflow-package}, e.g.[tests/providers/docker/operators/test_docker.py](https://github.com/apache/airflow/blob/main/tests/providers/docker/operators/test_docker.py)

Test environ variables might be done by [`monkeypatch`](https://docs.pytest.org/en/latest/how-to/monkeypatch.html) pytest fixture, something like that (just an example for the reference, 

```python
class TestDockerOperator:
    ...

    def test_respect_docker_host_env(self, monkeypatch):
        monkeypatch.setenv("DOCKER_HOST", "foo.bar")
        assert DockerOperator(task_id="test-task", image="test").docker_url == "foo.bar"

    def test_docker_host_env_empty(self, monkeypatch):
        monkeypatch.setenv("DOCKER_HOST", "")
        assert DockerOperator(task_id="test-task", image="test").docker_url == "expected-host-here"

    def test_docker_host_env_not_set(self, monkeypatch):
        monkeypatch.delenv("DOCKER_HOST", raising=False)
        assert DockerOperator(task_id="test-task", image="test").docker_url == "expected-host-here"
```

> In case of backwards incompatible changes please leave a note in a newsfragment file, named

Is it backward incompatible change?

**Comment by maresb** [comment]
Thanks so much @Taragolis for the advice! I will look into it.

> Is it backward incompatible change?

It depends on your definition of "backward incompatible". The only breakage I can imagine is quite contrived:

* The user is running two Docker hosts, one on the default `unix://var/run/docker.sock` and the other on `tcp://docker-proxy:2375`. The user intends to use `unix://var/run/docker.sock` with Airflow and `tcp://docker-proxy:2375` with everything else, so they set `DOCKER_HOST=tcp://docker-proxy:2375` and rely otherwise on the default value of `docker_url` for Airflow.
* The fix in this case is that, after this PR has been merged, the user must explicitly set `docker_url="unix://var/run/docker.sock"` in order to override the value in `DOCKER_HOST`.

Do you think this warrants a newsfragment?

**Comment by Taragolis** [comment]
> Do you think this warrants a newsfragment?

We do not use newfragments for the providers changes, so this not this case.

However if you thought it might break someone pipeline than better create a new major version of the provider.
In this case you have to add new record to the `provider.yaml` versions, `4.0.0` in this case
https://github.com/apache/airflow/blob/83316b81584c9e516a8142778fc509f19d95cc3e/airflow/providers/docker/provider.yaml#L27-L28 

And add into the [CHANGELOG.rst](https://github.com/apache/airflow/blob/83316b81584c9e516a8142778fc509f19d95cc3e/airflow/providers/docker/CHANGELOG.rst) info about Breaking changes and how it might be resolved, and what user should change

Some example for reference: https://github.com/apache/airflow/blob/bfb054e9e867b8b9a6a449e43bfba97f645e025e/airflow/providers/slack/CHANGELOG.rst?plain=1#L153-L188

---

cc: @eladkal @potiuk do you think this changes might considered as breaking changes?

**Comment by eladkal** [comment]
> > cc: @eladkal @potiuk do you think this changes might considered as breaking changes?
> 
> I'd treat is a bug-fix but add a separate note in the changelog. DOCKER_HOST overrididng default value is well-respected standard and we should follow it. It's very unlikely that somoene sets DOCKER_HOST and does not want it to be used by default.

I agree

File: airflow/decorators/__init__.pyi
Comment (Taragolis) [inline airflow/decorators/__init__.pyi]: I don't think we should add expressions into the stub file. It is not evaluated in runtime.

**Comment by maresb** [comment]
Thanks so much for the support!

I pushed some draft tests. I haven't set up testing locally, and I'm hoping that things are simple enough that I can use the CI test suite. Seems like I require approval, so I'll look into setting it up locally.

> I'd treat is a bug-fix but add a separate note in the changelog.

It looks like the changelog is managed semi-automatically, so I'm not sure if I should edit the [changelog](https://github.com/maresb/airflow/blob/docker-operator-docker-host-envvar/airflow/providers/docker/CHANGELOG.rst) by hand. If you want a one-liner, I could suggest:

> Use the `DOCKER_HOST` environment variable as the default value for the `docker_url` parameter.

File: airflow/providers/docker/operators/docker.py
Comment (Taragolis) [inline airflow/providers/docker/operators/docker.py]: Small nit

Just import `os` an use as `os.environ` as more convenient way

**Comment by potiuk** [comment]
> It looks like the changelog is managed semi-automatically

Add a one liner at the top please (see comment at the beginning of the changelog) -we will almost for sure forget to add the line if it is not added now and release manager will have to remember about looking back to all the PRs to see if one is missing.

File: airflow/decorators/__init__.pyi
Comment (Taragolis) [inline airflow/decorators/__init__.pyi:1]: Let's add changes to stub file into the separate PR, when this one is merged.
This change to the Airflow Core, and could make additional headache to release manager to determine in which version of Airflow it should be included

Comment (maresb) [reply airflow/decorators/__init__.pyi:1]: Done in https://github.com/apache/airflow/pull/38406

**Comment by maresb** [comment]
I think I've addressed all the feedback, and I'm optimistic that the tests are succeeding. I'll keep an eye on them and wait for everything to go green unless I hear more feedback. Thanks so much for the awesome review process!!!

File: airflow/providers/docker/CHANGELOG.rst
Comment (potiuk) [inline airflow/providers/docker/CHANGELOG.rst]: By note here i meant not the title of the PR (this will indeed be automatically retrieved) but a hand-written note on what to expect if you are user, explaining potential problems and remediations. Something like 

https://github.com/apache/airflow/blob/main/airflow/providers/google/CHANGELOG.rst#L113 for example

Comment (maresb) [reply airflow/providers/docker/CHANGELOG.rst]: Oops, I misunderstood; thanks for your patience. I updated it.

All CI checks that weren't skipped were green. :rocket:

**Review by potiuk — APPROVED** [review]
LGTM. @Taragolis @eladkal ?

**Review by Taragolis — APPROVED** [review]
Nice

**Pull request merged**
